Integrate bullet into ign-physics3

Blast545 opened the pull request:
This PR includes bullet integration into ign-physics, supporting the set of basic features described here:

The following elements are available:

* Creating basic shapes and meshes, with their associated collisions
* Step simulation
* Revolute joints with their getter/setters
* Diff drive plugin integration
* Fixed joints

Unit tests are not included in this PR, the elements in this PR were tested with the following example from gazebo4:
`shapes.sdf`, `joint_controller.sdf`, `joint_position_controller.sdf`. Meshes were tested on their own with #178.
Revolute joints and ground vehicles were tested with a modified version of `diff_drive.sdf` (replacing the ball joint with a fixed one). And using the subt simulation with the model X1_SENSOR_CONFIG_1. This test was run with the command:

`ign launch -v4 cave_circuit.ign circuit:=cave worldName:=simple_cave_01 robotName1:=X1 robotConfig1:=X1_SENSOR_CONFIG_1 engine:=bullet`

And using this subt fork: `https://github.com/Lobotuerk/subt/tree/bullet_test`.

## Known issues:

* There is an issue with the axis of some revolute joint combinations, this was found running this command:

![gimball_problem_](https://user-images.githubusercontent.com/8069967/106797407-4fa13280-663b-11eb-9b68-3ddb128ca09a.gif)

* The angle getter from bullet resets its initial count if the angle difference between two consecutive readings is higher than a hardcoded threshold. To address this issue in the joint controllers a threshold in the max torque was set. (An extra eyes into the best approach to address this would be appreciated). Link to the [threshold applied in bullet ](https://pybullet.org/Bullet/BulletFull/btHingeConstraint_8cpp_source.html#l00318) Im talking about.

* Fixed joints are not fixing all the possible degrees of freedom, @Lobotuerk ran into this testing flying vehicles.

Blast545 commented:
With 6742e5d it was added the changes proposed in `bulletDev/lobotuerk/ParameterTuning` and an extra variable `mu3` used for testing friction of the models.

Lobotuerk commented on examples/hello_world_loader/CMakeLists.txt:
```diff
@@ -3,7 +3,7 @@ cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
 find_package(ignition-plugin${IGN_PLUGIN_VER} 1.1 REQUIRED COMPONENTS all)
 
-set(IGN_PHYSICS_VER 3)
+set(IGN_PHYSICS_VER 4)
```
Why do we change this into 4, if we are integrating it into 3?

Lobotuerk commented on examples/hello_world_plugin/CMakeLists.txt:
```diff
@@ -3,7 +3,7 @@ cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
 find_package(ignition-plugin${IGN_PLUGIN_VER} 1.1 REQUIRED COMPONENTS all)
 
-set(IGN_PHYSICS_VER 3)
+set(IGN_PHYSICS_VER 4)
```
Same as before

Lobotuerk commented on bullet/src/SDFFeatures.cc:
```diff
@@ -0,0 +1,382 @@
+                              ->GetElement("ode");
+  const auto mu = odeFriction->Get<btScalar>("mu");
+  const auto mu2 = odeFriction->Get<btScalar>("mu2");
+  const auto mu3 = odeFriction->Get<btScalar>("mu3");
```
SDF format does not have a default value for mu3, so you get 0 here if not defined on the .sdf. Having 0 on one of the forces axis does not seem good (and tests probe that the whole movement of the robot misbehaves with this)

nkoenig replied:
Is this still an issue? If it is, you could first check if `mu3` is valid.

mjcarroll commented on bullet/src/Base.hh:
```diff
@@ -0,0 +1,214 @@
+    return this->GenerateIdentity(id, this->joints.at(id));
+  }
+
+  public: using WorldInfoPtr = std::shared_ptr<WorldInfo>;
+  public: using ModelInfoPtr = std::shared_ptr<ModelInfo>;
+  public: using LinkInfoPtr  = std::shared_ptr<LinkInfo>;
+  public: using CollisionInfoPtr = std::shared_ptr<CollisionInfo>;
+  public: using JointInfoPtr  = std::shared_ptr<JointInfo>;
```
Any reason for `shared` vs `unique` here?

Blast545 replied:
Those in particular can't be changed because the `GenerateIdentity` templated function from ignition takes a shared pointer to generate new identities.

mjcarroll commented on bullet/src/Base.hh:
```diff
@@ -0,0 +1,214 @@
+  public: std::unordered_map<std::size_t, JointInfoPtr> joints;
+  public: std::unordered_map<std::size_t, std::size_t> collisionGroups;
+
+  public: int internalTicksDivider = 0;
```
Unused?

Blast545 replied:
Removed with 363889c

mjcarroll commented on bullet/src/Base.hh:
```diff
@@ -0,0 +1,214 @@
+
+};
+
+}
+}
+}
```
```suggestion
}  // namespace bullet
}  // namespace physics
}  // namespace ignition
```

Blast545 replied:
Added all missing with ab71031

mjcarroll commented on bullet/src/EntityManagementFeatures.cc:
```diff
@@ -0,0 +1,186 @@
+  }
+
+  // Clean up links
+  std::unordered_map<std::size_t, LinkInfoPtr>::iterator it =
+    this->links.begin();
```
```suggestion
auto it = this->links.begin();
```

Blast545 replied:
Changed all similar iterators with 607b609

mjcarroll reviewed (commented):
Still reviewing, but so far I see a lot of manual memory management with raw pointers.  I think we should generally avoid this in new code unless there is a very specific reason.  Is there one in this case?

mjcarroll commented on bullet/src/JointFeatures.cc:
```diff
@@ -0,0 +1,456 @@
+    const JointInfoPtr &jointInfo = this->joints.at(_id.id);
+    const int jointType = jointInfo->constraintType;
+    // Check the type of joint and act accordignly
+    if (jointInfo->constraintType ==
```
Is the intent to handle other constraints?  In that case I think a switch makes more sense here.

Blast545 replied:
Yes, I changed it to a switch statement with 7d6d9d2

Blast545 commented:
@mjcarroll There is no specific reason to use raw pointers, thanks for the comments, I'll address them today.

Blast545 commented:
Added some extra commits:

> c583765 , 4cf6b4e , aa22784 These commits were added to set and test mu and mu2 friction values as close as possible to DART physics implementation. With the worlds added `./bullet/worlds/test_friction_mu.sdf` and `./bullet/worlds/test_friction_mu2.sdf` it can be seen the axis where those friction components are applied.

> d022c34 This commit was added to remove collision groups from this PR. Links within a model will collide with each other if they overlap only if they are not sharing a joint. However, that's not a common scenario.

> 6c1dbd6 ae99a36 These commits were added to remove/update outdated comments in the code. PTAL @Lobotuerk 

> 9f332c9 Remove joints from bullet world when a model is removed, PTAL @Lobotuerk 

> dc325c0 Divide inertia from torque to get acceleration. Can someone with more physics experience help me out figuring out if this makes sense?

Blast545 commented on bullet/src/Base.hh:
```diff
@@ -0,0 +1,214 @@
+///    GenerateIdentity.
+/// 3) Hold explicit copies of raw pointers that can be deallocated
+
+// todo(anyone): Handle cleaning these pointers
```
I think the same as you, I think those can be `std::unique_ptrs` as well, but I implemented those with `std::shared_ptrs` to copy the same pattern used with `dart` implementation. I have in this commit in a test branch: [Smart pointers](https://github.com/ignitionrobotics/ign-physics/commit/f7b1d0af136feedb584ef28acc4085a9fcd286c9). 

However, when I add this, the program crashes when it's closed, showing a `malloc_consolidate` error. or a `free pointer, invalid chunk size`. I think bullet has problem when the things are not deleted in reverse order and there is no guarantee with ignition on its own.

Blast545 replied:
Never mind, I was able to use `shared_ptrs` in this PR, take a look to `bd154bb`. With this I think we've addressed your comments, let me know if this is OK

Blast545 commented:
@Lobotuerk As this branch does not include smart pointers for memory management (at least yet) I added the missing delete with 607b609

Blast545 commented:
# Current status
The following elements are available:

- Creating basic shapes and meshes, with their associated collisions
- Step simulation
- Revolute joints with their getter/setters
- Diff drive plugin integration
- joint_controller and joint_controller_position integration
- Fixed joints
- Basic friction model (mu + mu2)

## To-Do: items
-   Flying vehicles: Required features implemented on top of `bulletDev/lobotuerk/UAV`, this effort was dropped to make sure UGV’s were working properly.
-   Joints damping: Partially implemented on top of `bulletDev/blast545/bulletPhysicsPointers`. As it is not fully tested, is not included in this PR.
-   Replace raw pointers with smart pointers: Partially implemented on top of `bulletDev/blast545/bulletPhysicsPointers`. There seems to be a problem with how the std library manages the bullet elements so using smart pointers is causing problems. Not included in this PR at the moment.
-   Other types of joints
-   Detachable joints
-   Canonical links
-   Contacts getter
-   Self Collisions / collision groups (self_collide sdf tag): This was partially implemented and removed with commit: <link to commit>. Dropped because it had limitations and didn’t add value to the features available.
-   Move things on GUI
-   Enable movement thresholds to disable objects not moving
-  Unit testing
-   Memory management
-   Levels

## Known issues

Issue with inertia diagonal values: When small inertia values are used with bullet objects, they cause erratic behaviors with the simulation. This issue was found when testing models with gimbals, an inertia matrix with low inertia values is present in a cylinder inside the gimbal assembly. A cylinder equivalent to the one found, with same inertia matrix, was tested using gazebo11, ign-gazebo4 and the bullet libraries standalone and it fails. However, this works fine with pyBullet, so @Lobotuerk  is looking into that ATM.

![Test_weird_result](https://user-images.githubusercontent.com/8069967/110865843-016af900-82a3-11eb-96d1-1bb9320efe6a.gif)


Issue with joints: Update here: Damping functionality was added and seems to be working like it should. The only time where it doesn’t work is with the weird inertial matrix.

Torque limitation: * The angle getter from bullet resets its initial count if the angle difference between two consecutive readings is higher than a hardcoded threshold. To address this issue in the joint controllers a threshold in the max torque was set. (An extra eyes into the best approach to address this would be appreciated). This is the code we implemented <<https://github.com/ignitionrobotics/ign-physics/blob/af978795a077165d91067d11c10912b011b50d4b/bullet/src/JointFeatures.cc#L291>> and this is the limitation addressed <<https://pybullet.org/Bullet/BulletFull/btHingeConstraint_8cpp_source.html#l00318>>, also @Lobotuerk is taking a look into Gazebo11 implementation this week.

## Tests available

- From gazebo4, to test basic features:
`shapes.sdf`, `joint_controller.sdf`, `joint_position_controller.sdf`. 
- To test inertia related issues: `test_gimball_base_cylinder.sdf`, `test_gimball_joint.sdf`, `test_hurskey_gimball_model.sdf`
- To test mu,mu2 friction directions: `test_friction_mu2.sdf`, `test_friction_mu.sdf`
- To test effect default behavior of bullet for overlapping objects: , `test_x1_without_joints.sdf`,  `test_hurskey_without_joints.sdf`
- To test friction + collisions in a general way: `test_shapes_gravity_pose.sdf`, `test_friction_multiple.sdf`, `test_friction_single.sdf  `
- To test diff drive plugin + complete model: `test_hurskey_vehicle.sdf`, `test_x1.sdf`, `test_x1_tunnel.sdf`

Frictions test `test_friction_mu.sdf`
![test_friction_mu_bullet_modified](https://user-images.githubusercontent.com/8069967/110865956-2fe8d400-82a3-11eb-8556-a0e1bff5aca0.gif)


Test diff_drive vehicle: `test_x1_tunnel.sdf`

![test_x1_tunnel (1)](https://user-images.githubusercontent.com/8069967/110866166-7c341400-82a3-11eb-81dd-17d5b19c488e.gif)

azeey commented on bullet/src/SDFFeatures.cc:
```diff
@@ -0,0 +1,368 @@
+  const math::Pose3d pose = _sdfLink.RawPose();
+  const ignition::math::Inertiald inertial = _sdfLink.Inertial();
+  auto mass = inertial.MassMatrix().Mass();
+  const auto diagonalMoments = inertial.MassMatrix().DiagonalMoments();
```
This only takes the diagonal moments the mass matrix and ignores the off diagonal terms. Based on the Gazebo-classic implementation, If I understand correctly, when constructing `btRigidBody`, we have to give bullet 3 quantities: 
 1. the position of the body frame. The origin of this frame is a the center of mass of the body.
 2. the orientation of the body frame such that the inertia matrix is diagonal. This means we have to diagonalize the inertia matrix and apply the resulting rotation offset to the pose of the body frame.
 3. the diagonal moments of the diagonalized inertia matrix.

To accomplish (1), we have to compose the model pose, the link pose and the inertia pose. I think `poseIsometry` is missing `inertial.Pose()`.

For (2), we have to first use `ignition::math::MassMatrix3::PrincipalMoments` which diagonalizes the inertia matrix. `ignition::math::MassMatrix3::DiagonalMoments` simply gives the values on the diagonal of the matrix.

After that, we have to obtain the rotation offset using `PrincipalAxesOffset` and apply that to `inertial.Pose()`.

Lobotuerk replied:
https://github.com/ignitionrobotics/ign-physics/pull/226 should take care of this. @azeey If you can think of an easy way to test it let me know. I can't wrap my head around the usage of the inertial pose yet

Blast545 commented on bullet/src/SDFFeatures.cc:
```diff
@@ -0,0 +1,368 @@
+
+  auto gravity = _sdfWorld.Gravity();
+  worldInfo->world->setGravity(btVector3(gravity[0], gravity[1], gravity[2]));
+  return worldID;
```
You are right, we'll add it

Blast545 replied:
@azeey We saw this is included in DART and TPE, but it seems the SDF parser already iterate over the structures internally, any reason to add the links here if the code will already call AddLink on each link? Maybe we are missing something

azeey replied:
> it seems the SDF parser already iterate over the structures internally

Which SDF parser are you referring to? 

The code in ign-gazebo calls `ConstructModel` for each model, but that's just one usage of ign-physics. One reason it's done that way in ign-gazebo is to support levels. When ign-gazebo calls `ConstructWorld`, `_sdfWorld` won't have any models. But the expectation is that `ConstructWorld` would construct the entire world, including any models it finds in `_sdfWorld`. For example, the tests in `dartsim/SimulationFeatures_TEST` call `ConstructWorld`, but never call `ConstructModel` themselves.

Blast545 replied:
I didn't think about a different case scenario for ign-physics, I'll add this

Blast545 replied:
Addressed with 829e120, PTAL

Blast545 commented on bullet/src/SimulationFeatures.hh:
```diff
@@ -0,0 +1,59 @@
+      const Identity &/* _worldID */) const override
+      {
+        // TODO(lobotuerk): Implement contacts getter, could be like https://pybullet.org/Bullet/phpBB3/viewtopic.php?t=2855
+        ignerr << "Dummy GetContactsFromLastStep implementation";
```
IIRC @Lobotuerk Added them to be able of adding collisions in `ign-gazebo`, the minimun set of features needed for collisions included this one [Collision features](https://github.com/ignitionrobotics/ign-gazebo/blob/d79aafd9de25c4d4b88bc210be3faefd40214387/src/systems/physics/Physics.cc#L303). So if this feature is not implemented, collisions are skipped.

azeey replied:
Okay. It's just that the whole idea of the `Feature` system in ign-physics is to avoid dummy implementations. If there's a plan to implement this in the short term, I'm good with merging this as is. If it's not going to be implemented, maybe we can look into making the `GetContactsFromLastStepFeature` feature an optional feature.

Lobotuerk replied:
https://github.com/ignitionrobotics/ign-gazebo/commit/a84ec51c2fddcc09eeb0ed26b31f8de669c1ce56 should resolve this

Blast545 commented on bullet/src/SDFFeatures.cc:
```diff
@@ -0,0 +1,368 @@
+{
+  // Read sdf params
+  const std::string name = _sdfModel.Name();
+  const math::Pose3d pose = _sdfModel.RawPose();
```
@azeey You are right, I addressed this with edf9e8e. I actually used a test case sdf to check the differences but didn't see any change in the result. Am I missing something? The results let me to think RawPose is giving me the resolved poses already.

azeey replied:
When ign-gazebo sends poses, they should always be resolved. That is done as a work around to the limitation of the communication between ign-gazebo and ign-physics. Users that use ign-physics directly to load SDFormat files could have poses that are not resolved.

azeey commented on bullet/src/EntityManagementFeatures.cc:
```diff
@@ -0,0 +1,213 @@
+    const Identity & _worldID, std::size_t _modelIndex)
+{
+  // Check if the model exists
+  if (this->models.find(_modelIndex) == this->models.end() ||
```
Wouldn't this check against model IDs instead of indices?

Blast545 replied:
Yeah, we are using the generated entities values as indices to save the objects inside unordered maps. So, there is an equivalence modelID.id to index. Would this cause a problem? Maybe we are missing something

azeey replied:
Well, it depends on the way the IDs are generated. Right now, the ID is incremented for any entity created. Consider this SDF
```xml
<sdf version='1.6'>
  <world name='W'>
    <model name='M0'>
      <link name='L'/>
    </model>
   <model name='M1'/>
  </world>
</sdf>
```

The index of model `M1` in the world should always be 1, but if the order of creation happens to be `W`, `M0`, `L`, `M1`, the ID of `M1` would be 3!

This is also problematic when models are removed and created at run time. The ID would always increase, but the index may not.

Blast545 replied:
>> The index of model M1 in the world should always be 1

Why? Is this information saved/generated elsewhere? Like something checking how many models are in each world?

azeey replied:
Yeah, if you want to loop through all the models, you can do a for loop starting from 0 to `World::ModelCount()`.

Blast545 replied:
Should be addressed with #240, PTAL

azeey commented on bullet/src/EntityManagementFeatures.cc:
```diff
@@ -0,0 +1,213 @@
+    return false;
+  }
+
+  auto model = this->models.at(_modelIndex);
```
I think this can be refactored into it's own function because `RemoveModel` is doing the same thing.

Blast545 replied:
You are right, fff1f08. This also considers index <-> ID.id

azeey commented on bullet/src/EntityManagementFeatures.cc:
```diff
@@ -0,0 +1,213 @@
+    if (modelInfo->name == _modelName)
+    {
+      found = true;
+      index_id = model.first;
```
I think this is the model ID, not the index.

Blast545 commented on bullet/src/JointFeatures.cc:
```diff
@@ -0,0 +1,478 @@
+      case static_cast<int>(::sdf::JointType::REVOLUTE) :
+      {
+        btHingeAccumulatedAngleConstraint* hinge =
+	  static_cast<btHingeAccumulatedAngleConstraint*>(jointInfo->joint.get());
```
Changed with 04c3f21

Blast545 commented on bullet/src/JointFeatures.cc:
```diff
@@ -0,0 +1,478 @@
+      case static_cast<int>(::sdf::JointType::REVOLUTE) :
+      {
+	btHingeAccumulatedAngleConstraint* hinge =
+	  static_cast<btHingeAccumulatedAngleConstraint*>(jointInfo->joint.get());
```
Changed with 04c3f21

Blast545 commented on bullet/src/JointFeatures.cc:
```diff
@@ -0,0 +1,478 @@
+	  // angular position of the joint and losing the angle reference
+	  // TO-DO (blast545): this limitation should be based on angular speed
+	  // as this breaks the PID controller when setting high values
+	  const double thresholdValue = std::max(std::min(_value, 0.1), -0.1);
```
@azeey Mmm, we added this because increasing the speed of a revolute joint up to a certain point made the internal controller to lose track of the angle. We noted it as a known issue, but IIRC it only affected aerial vehicles. I'm OK with removing the threshold as long as the "known issue" remains visible. WDYT?

@Lobotuerk Also WDYT

Lobotuerk replied:
I agree, lets remove the threshold, and point out in the issue to the bullet's codebase part that has that "if speed > cost, free angle"

azeey replied:
Which controller are you referring to?

Blast545 replied:
It's not a controller, it's the internal "angle tracker" that bullet uses. You can see the problem we are refering to here: https://pybullet.org/Bullet/BulletFull/btHingeConstraint_8cpp_source.html#l00318. If there's a change in the angle bigger than the internal threshold within two simulation times, the angle tracker resets the count to the absolute current angle.

Blast545 commented on bullet/src/JointFeatures.cc:
```diff
@@ -0,0 +1,478 @@
+  ignwarn << "Dummy SetRevoluteJointAxis\n";
+}
+
+/////////////////////////////////////////////////
```
Removed with 0b4b940

Blast545 commented on bullet/src/JointFeatures.cc:
```diff
@@ -0,0 +1,478 @@
+}
+
+/////////////////////////////////////////////////
+Identity JointFeatures::AttachFixedJoint(
```
Removed 0b4b940

Blast545 commented on bullet/src/JointFeatures.cc:
```diff
@@ -0,0 +1,478 @@
+    const Identity &_jointID, const AngularVector3d &_axis)
+{
+  (void) _jointID;
+  (void) _axis;
```
I removed this setter from the scope of the PR as it was a dummy implementation not needed, 0b4b940

Blast545 commented on bullet/src/JointFeatures.cc:
```diff
@@ -0,0 +1,478 @@
+  (void) _id;
+  // Degrees of freedom may need to be saved in the JointInfo struct
+  // Currently supporting 1DoF joints
+  return 1;
```
You are right, fixed with 13931a5

Blast545 commented on bullet/src/JointFeatures.cc:
```diff
@@ -0,0 +1,478 @@
+  const Identity &_jointID) const
+{
+  (void) _jointID;
+  ignwarn << "Dummy CastToFixedJoint\n";
```
You are right, fixed with 13931a5. Also included an equivalent for `CastToRevoluteJoint`

Blast545 commented on bullet/src/JointFeatures.cc:
```diff
@@ -0,0 +1,478 @@
+AngularVector3d JointFeatures::GetRevoluteJointAxis(
+    const Identity &_jointID) const
+{
+  (void) _jointID;
```
I added this with 13931a5. I'm not sure if the angularVector consumed here is the same one used for the `GetJointVelocity`, let me know if I'm missing something

azeey replied:
It's not clear in the documentation of the `GetRevoluteJointProperties` feature, but from what's implemented in dartsim-plugin, I think we want the axis in the joint frame, not in the world frame. I'm not 100% sure, but this code returns the axis in world frame. I think we can just premultipy it with the inverse of the joint's rotation in world.

Blast545 commented on bullet/src/ShapeFeatures.cc:
```diff
@@ -0,0 +1,113 @@
+    mTriMesh.get()->addTriangle(bv0, bv1, bv2);
+  }
+
+  auto gimpactMeshShape =
```
This creates a `btGImpactMeshShape` on top of a `btTriangleMesh`, is not a copy. Also, you can't destroy the reference to `btTriangleMesh` either, because the first depends on the second one x_x

azeey replied:
Ah, I missed this is a different type.

Blast545 commented on bullet/src/ShapeFeatures.cc:
```diff
@@ -0,0 +1,113 @@
+  unsigned int numVertices = _mesh.VertexCount();
+  unsigned int numIndices = _mesh.IndexCount();
+
+  const auto mTriMesh = std::make_shared<btTriangleMesh>(new btTriangleMesh());
```
You are right, thanks! dcaee86

Blast545 commented on bullet/src/KinematicsFeatures.cc:
```diff
@@ -0,0 +1,81 @@
+    return data;
+  }
+  const auto &linkInfo = this->links.at(linkID);
+  const auto &model = linkInfo->link;
```
Also that's not a model, renamed with 7324eb4

Blast545 commented on bullet/src/ShapeFeatures.cc:
```diff
@@ -0,0 +1,113 @@
+Identity ShapeFeatures::CastToMeshShape(
+    const Identity &_shapeID) const
+{
+  ignwarn << "Dummy implementation CastToMeshShape" << std::endl;
```
Is not clear to me how to implement this, as we are saving the collision objects, not the shapes itself. Inside the `CollisionInfo` objects we have a `std::shared_ptr<btCollisionShape> shape;` but I dont think this is the mesh shape expected by ignition. 

And, if we can return this one, do we have to generate a new identity for it? or use the same identity value as the one that the collision object has?

Lobotuerk replied:
Maybe we can add a bool to collisionInfo and check if that bool is true in this function?

Lobotuerk replied:
Added in 6df5d27

Blast545 commented on bullet/src/SDFFeatures.cc:
```diff
@@ -0,0 +1,456 @@
+  else
+  {
+    axis = (ResolveSdfPose(_sdfJoint.SemanticPose()) + this->links.at(childId)->pose).Rot()
+       * _sdfJoint.Axis(0)->Xyz();
```
Addressed, see 7121887

Lobotuerk commented on bullet/src/KinematicsFeatures.cc:
```diff
@@ -0,0 +1,81 @@
+  const auto &model = linkInfo->link;
+
+  btTransform trans;
+  model->getMotionState()->getWorldTransform(trans);
```
From what I have learned on forums and places like such, the only difference between getting the transformation from the motionState vs model GetCenterOfMassTransform is that GetCenterOfMassTransform is actual without taking into account forces, and taking it from motionState interpolates into the future. I will change it, but from practical results I tried both and it does not change anything.

Lobotuerk replied:
Updated on 6df5d27

Lobotuerk commented on bullet/src/SDFFeatures.cc:
```diff
@@ -0,0 +1,456 @@
+    const auto poseLinear = poseIsometry.linear();
+    btTransform baseTransform;
+    baseTransform.setOrigin(convertVec(poseTranslation));
+    baseTransform.setBasis(convertMat(poseLinear));
```
There is a lot of conversions functions we could use, but I don't think we have the time for all of them

Lobotuerk commented on bullet/src/SDFFeatures.cc:
```diff
@@ -0,0 +1,456 @@
+  }
+  else
+  {
+    axis = (ResolveSdfPose(_sdfJoint.SemanticPose()) + this->links.at(childId)->pose).Rot()
```
Updated on 6df5d27

Lobotuerk commented on bullet/src/KinematicsFeatures.cc:
```diff
@@ -0,0 +1,81 @@
+  // omega = matBaseToWorld * omega;
+  // vel = matBaseToWorld * vel;
+
+  data.linearVelocity = convert(vel);
```
Please check dffee9f to see if I understood you correctly

azeey replied:
Looking at [code](https://github.com/osrf/gazebo/blob/ea72e4d624ffef409694c6624fffd43b6a946667/gazebo/physics/bullet/BulletLink.cc#L415-L434) in Gazebo classic, I think we need to rotate `inertialPose.Pos()` into the world frame because both `vel` and `omega` are given in the world frame at this point.

The code there actually uses[ `btRigidBody::getVelocityInLocalPoint`](https://github.com/bulletphysics/bullet3/blob/master/src/BulletDynamics/Dynamics/btRigidBody.h#L460-L463), but that is simply doing the cross product as well.

Lobotuerk replied:
I dont have a proper way to test this, but 024bdb2 should take care, at least the math seems to check

azeey replied:
This changed the order of operands in the cross product. I'm not sure that's correct. The formula is usually given as `omega x p` where `p` is the vector from the COM to the point of interest. But in this case `inertiaPose.Pos()` is the vector from the link origin to COM, so we should either reverse the order of operands or add a `-` sign to `inertiaPose.Pos()`.

Lobotuerk replied:
040bc9e should fix this, thank you @azeey for the help

azeey commented on bullet/src/Base.hh:
```diff
@@ -0,0 +1,257 @@
+    auto it = this->childIdToParentId.find(_id);
+    if (it != this->childIdToParentId.end())
+    {
+      for (const auto &pair : this->childIdToParentId)
```
This works because `childIdToParentId` is an ordered map and so this iteration proceeds in ascending order of the keys of that map. Can you add a comment to this variable's declaration so someone doesn't change it to `unordered_map` sometime down the road?

Lobotuerk replied:
Added the comment in 024bdb2

chapulina commented on bullet/src/SimulationFeatures.cc:
```diff
@@ -0,0 +1,40 @@
+    ForwardStep::State & /*_x*/,
+    const ForwardStep::Input & _u)
+{
+    const WorldInfoPtr &worldInfo = this->worlds.at(_worldID);
```
Should we check if the world pointer is valid before using it below?

Blast545 replied:
Mmm, AFAIK there's no way this function can get a wrong `_worldID` because the upper layers operate over registered `worldIDs`. Mind correcting me if I am wrong.

@azeey wdyt

azeey replied:
I remember a discussion between @scpeters  and @mxgrey that if the ID is not in the container, something has gone terribly wrong and so an exception would make sense. So, a lot of places in ign-physics follow the same pattern of calling `at`, even though it can throw.

chapulina commented on bullet/src/ShapeFeatures.cc:
```diff
@@ -0,0 +1,120 @@
+Identity ShapeFeatures::CastToMeshShape(
+    const Identity &_shapeID) const
+{
+  if (this->collisions.at(_shapeID)->isMesh)
```
Should we check that there's a valid pointer there before calling `isMesh`?

Lobotuerk replied:
`unordered_map.at()` does not return a non valid pointer AFAIK, its a valid or it throws and exception

Lobotuerk replied:
https://github.com/ignitionrobotics/ign-physics/pull/208#discussion_r623903046 This applies here too

chapulina reviewed (commented):
Just took a high-level view and left some comments.

Blast545 commented on bullet/src/SDFFeatures.cc:
```diff
@@ -0,0 +1,480 @@
+    convertVec(ignition::math::eigen3::convert(diagonalMoments));
+
+  const auto &modelInfo = this->models.at(_modelID);
+  math::Pose3d base_pose = modelInfo->pose;
```
Addressed with: c3e197d

Blast545 commented on bullet/CMakeLists.txt:
```diff
@@ -0,0 +1,68 @@
+EXECUTE_PROCESS(COMMAND ${CMAKE_COMMAND} -E create_symlink ${versioned} ${unversioned})
+INSTALL(FILES ${PROJECT_BINARY_DIR}/${unversioned} DESTINATION ${IGNITION_PHYSICS_ENGINE_INSTALL_DIR})
+
+# Testing
```
Moved some basic from DART 46e1d22. We followed your advice of moving DART's tests to bullet's but a lot of them rely on `GetEntities` and `ConstructEmptyModel` features that this PR doesn't support :/

Blast545 commented on bullet/src/JointFeatures.hh:
```diff
@@ -0,0 +1,121 @@
+{
+  // ----- Get Basic Joint State -----
+  public: double GetJointPosition(
+      const Identity &_id, const std::size_t _dof) const;
```
Addressed: 6f5bdb4

Blast545 commented on bullet/src/EntityManagementFeatures.hh:
```diff
@@ -0,0 +1,65 @@
+
+  public: bool RemoveModelByName(
+      const Identity &/* _worldID */,
+      const std::string &/* _modelName */) override;
```
Addressed: f86bae6

Blast545 commented on bullet/worlds/meshes/VLP16_base_1.dae:
```diff
@@ -0,0 +1,89 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no" ?>
```
We added those to make easier for someone checking this branch to reproduce some results / issues with the code. Do you think we can post these world elsewhere in particular? Otherwise I can just move them to my personal repo and add a link somewhere in a readme.

nkoenig commented:
I'm seeing the following compiler error:

```
In file included from /home/nkoenig/work/ignition/dome/ign-physics/include/ignition/physics/FeatureList.hh:112,
                 from /home/nkoenig/work/ignition/dome/ign-physics/include/ignition/physics/RequestEngine.hh:25,
                 from /home/nkoenig/work/ignition/dome/ign-physics/bullet/src/EntityManagement_TEST.cc:24:
/home/nkoenig/work/ignition/dome/ign-physics/include/ignition/physics/detail/FeatureList.hh: In instantiation of ‘struct ignition::physics::detail::ComposePlugin<ignition::physics::FeaturePolicy<double, 3>, TestFeatureList, void>::type’:
/home/nkoenig/local/include/ignition/plugin1/ignition/plugin/detail/SpecializedPlugin.hh:202:13:   required from ‘class ignition::plugin::detail::SelectSpecializers<ignition::physics::detail::ComposePlugin<ignition::physics::FeaturePolicy<double, 3>, TestFeatureList, void>::type>’
/home/nkoenig/work/ignition/dome/ign-physics/include/ignition/physics/detail/FeatureList.hh:127:16:   required from ‘struct ignition::physics::detail::DeterminePlugin<ignition::physics::FeaturePolicy<double, 3>, TestFeatureList>::Specializer’
/home/nkoenig/work/ignition/dome/ign-physics/include/ignition/physics/detail/RequestEngine.hh:67:12:   required from ‘static ignition::physics::RequestEngine<FeaturePolicyT, FeatureListT>::EnginePtrType ignition::physics::RequestEngine<FeaturePolicyT, FeatureListT>::From(const PtrT&, std::size_t) [with PtrT = ignition::plugin::TemplatePluginPtr<ignition::plugin::Plugin>; FeaturePolicyT = ignition::physics::FeaturePolicy<double, 3>; FeatureListT = TestFeatureList; ignition::physics::RequestEngine<FeaturePolicyT, FeatureListT>::EnginePtrType = ignition::physics::EntityPtr<ignition::physics::Engine<ignition::physics::FeaturePolicy<double, 3>, TestFeatureList> >; std::size_t = long unsigned int]’
/home/nkoenig/work/ignition/dome/ign-physics/bullet/src/EntityManagement_TEST.cc:41:71:   required from here
/home/nkoenig/work/ignition/dome/ign-physics/include/ignition/physics/detail/FeatureList.hh:109:16: warning: direct base ‘ignition::physics::detail::ComposePlugin<ignition::physics::FeaturePolicy<double, 3>, void, void>’ inaccessible in ‘ignition::physics::detail::ComposePlugin<ignition::physics::FeaturePolicy<double, 3>, TestFeatureList, void>::type’ due to ambiguity
         struct type :
                ^~~~
```

nkoenig commented:
Does anyone see the follow segfault when running `ign launch -v4 cave_circuit.ign circuit:=cave worldName:=simple_cave_01 robotName1:=X1 robotConfig1:=X1_SENSOR_CONFIG_1 engine:=bullet`?

```
Stack trace (most recent call last) in thread 5730:
#21   Object "[0xffffffffffffffff]", at 0xffffffffffffffff, in 
#20   Object "/lib/x86_64-linux-gnu/libc.so.6", at 0x7fbe7c9ab71e, in clone
#19   Object "/lib/x86_64-linux-gnu/libpthread.so.0", at 0x7fbe7c6726da, in 
#18   Object "/usr/lib/x86_64-linux-gnu/libstdc++.so.6", at 0x7fbe797016de, in 
#17   Object "/home/nkoenig/local/lib/libignition-gazebo4.so.4", at 0x7fbe6a0d1259, in 
#16   Object "/home/nkoenig/local/lib/libignition-gazebo4.so.4", at 0x7fbe6a0dbe3a, in ignition::gazebo::v4::SimulationRunner::Run(unsigned long)
#15   Object "/home/nkoenig/local/lib/libignition-gazebo4.so.4", at 0x7fbe6a0db736, in ignition::gazebo::v4::SimulationRunner::Step(ignition::gazebo::v4::UpdateInfo const&)
#14   Object "/home/nkoenig/local/lib/libignition-gazebo4.so.4", at 0x7fbe6a0d5ce1, in ignition::gazebo::v4::SimulationRunner::UpdateSystems()
#13   Object "/home/nkoenig/local/lib/ign-gazebo-4/plugins/libignition-gazebo-physics-system.so", at 0x7fbe447b2264, in ignition::gazebo::v4::systems::Physics::Update(ignition::gazebo::v4::UpdateInfo const&, ignition::gazebo::v4::EntityComponentManager&)
#12   Object "/home/nkoenig/local/lib/ign-gazebo-4/plugins/libignition-gazebo-physics-system.so", at 0x7fbe447a90eb, in ignition::gazebo::v4::systems::PhysicsPrivate::Step(std::chrono::duration<long, std::ratio<1l, 1000000000l> > const&)
#11   Object "/usr/lib/x86_64-linux-gnu/libBulletDynamics.so.2.87", at 0x7fbe43fcd8e4, in btDiscreteDynamicsWorld::stepSimulation(float, int, float)
#10   Object "/usr/lib/x86_64-linux-gnu/libBulletDynamics.so.2.87", at 0x7fbe43fcc361, in btDiscreteDynamicsWorld::internalSingleStepSimulation(float)
#9    Object "/usr/lib/x86_64-linux-gnu/libBulletCollision.so.2.87", at 0x7fbe43ccf95e, in btCollisionWorld::performDiscreteCollisionDetection()
#8    Object "/usr/lib/x86_64-linux-gnu/libBulletCollision.so.2.87", at 0x7fbe43ccba2f, in btCollisionDispatcher::dispatchAllCollisionPairs(btOverlappingPairCache*, btDispatcherInfo const&, btDispatcher*)
#7    Object "/usr/lib/x86_64-linux-gnu/libBulletCollision.so.2.87", at 0x7fbe43cbcbac, in btHashedOverlappingPairCache::processAllOverlappingPairs(btOverlapCallback*, btDispatcher*)
#6    Object "/usr/lib/x86_64-linux-gnu/libBulletCollision.so.2.87", at 0x7fbe43ccc517, in btCollisionPairCallback::processOverlap(btBroadphasePair&)
#5    Object "/usr/lib/x86_64-linux-gnu/libBulletCollision.so.2.87", at 0x7fbe43ccbc73, in btCollisionDispatcher::defaultNearCallback(btBroadphasePair&, btCollisionDispatcher&, btDispatcherInfo const&)
#4    Object "/usr/lib/x86_64-linux-gnu/libBulletCollision.so.2.87", at 0x7fbe43ce7136, in btCompoundCompoundCollisionAlgorithm::processCollision(btCollisionObjectWrapper const*, btCollisionObjectWrapper const*, btDispatcherInfo const&, btManifoldResult*)
#3    Object "/usr/lib/x86_64-linux-gnu/libBulletCollision.so.2.87", at 0x7fbe43ce7e6c, in btCompoundCompoundLeafCallback::Process(btDbvtNode const*, btDbvtNode const*)
#2    Object "/usr/lib/x86_64-linux-gnu/libBulletCollision.so.2.87", at 0x7fbe43d2d484, in btGImpactCollisionAlgorithm::gimpact_vs_shape(btCollisionObjectWrapper const*, btCollisionObjectWrapper const*, btGImpactShapeInterface const*, btCollisionShape const*, bool)
#1    Object "/usr/lib/x86_64-linux-gnu/libBulletCollision.so.2.87", at 0x7fbe43d2d5db, in btGImpactCollisionAlgorithm::gimpact_vs_shape(btCollisionObjectWrapper const*, btCollisionObjectWrapper const*, btGImpactShapeInterface const*, btCollisionShape const*, bool)
#0    Object "/usr/lib/x86_64-linux-gnu/libBulletCollision.so.2.87", at 0x7fbe43d31443, in btGImpactMeshShapePart::lockChildShapes() const
Segmentation fault (Address not mapped to object [0x3270696c93])
Segmentation fault (core dumped)
```

Lobotuerk commented on bullet/src/EntityManagementFeatures.hh:
```diff
@@ -0,0 +1,65 @@
+namespace physics {
+namespace bullet {
+
+using EntityManagementFeatureList = FeatureList<
```
0d050dd should address this one

Lobotuerk commented on bullet/CMakeLists.txt:
```diff
@@ -0,0 +1,68 @@
+set(unversioned ${CMAKE_SHARED_LIBRARY_PREFIX}${PROJECT_NAME_NO_VERSION_LOWER}-${engine_name}${CMAKE_SHARED_LIBRARY_SUFFIX})
+if (WIN32)
+  # create_symlink requires cmake 3.13 on windows
+  cmake_minimum_required(VERSION 3.13 FATAL_ERROR)
```
0d050dd should address this one

Lobotuerk commented:
> I'm seeing the following compiler error:
> 
> ```
> In file included from /home/nkoenig/work/ignition/dome/ign-physics/include/ignition/physics/FeatureList.hh:112,
>                  from /home/nkoenig/work/ignition/dome/ign-physics/include/ignition/physics/RequestEngine.hh:25,
>                  from /home/nkoenig/work/ignition/dome/ign-physics/bullet/src/EntityManagement_TEST.cc:24:
> /home/nkoenig/work/ignition/dome/ign-physics/include/ignition/physics/detail/FeatureList.hh: In instantiation of ‘struct ignition::physics::detail::ComposePlugin<ignition::physics::FeaturePolicy<double, 3>, TestFeatureList, void>::type’:
> /home/nkoenig/local/include/ignition/plugin1/ignition/plugin/detail/SpecializedPlugin.hh:202:13:   required from ‘class ignition::plugin::detail::SelectSpecializers<ignition::physics::detail::ComposePlugin<ignition::physics::FeaturePolicy<double, 3>, TestFeatureList, void>::type>’
> /home/nkoenig/work/ignition/dome/ign-physics/include/ignition/physics/detail/FeatureList.hh:127:16:   required from ‘struct ignition::physics::detail::DeterminePlugin<ignition::physics::FeaturePolicy<double, 3>, TestFeatureList>::Specializer’
> /home/nkoenig/work/ignition/dome/ign-physics/include/ignition/physics/detail/RequestEngine.hh:67:12:   required from ‘static ignition::physics::RequestEngine<FeaturePolicyT, FeatureListT>::EnginePtrType ignition::physics::RequestEngine<FeaturePolicyT, FeatureListT>::From(const PtrT&, std::size_t) [with PtrT = ignition::plugin::TemplatePluginPtr<ignition::plugin::Plugin>; FeaturePolicyT = ignition::physics::FeaturePolicy<double, 3>; FeatureListT = TestFeatureList; ignition::physics::RequestEngine<FeaturePolicyT, FeatureListT>::EnginePtrType = ignition::physics::EntityPtr<ignition::physics::Engine<ignition::physics::FeaturePolicy<double, 3>, TestFeatureList> >; std::size_t = long unsigned int]’
> /home/nkoenig/work/ignition/dome/ign-physics/bullet/src/EntityManagement_TEST.cc:41:71:   required from here
> /home/nkoenig/work/ignition/dome/ign-physics/include/ignition/physics/detail/FeatureList.hh:109:16: warning: direct base ‘ignition::physics::detail::ComposePlugin<ignition::physics::FeaturePolicy<double, 3>, void, void>’ inaccessible in ‘ignition::physics::detail::ComposePlugin<ignition::physics::FeaturePolicy<double, 3>, TestFeatureList, void>::type’ due to ambiguity
>          struct type :
>                 ^~~~
> ```
Fixed by a72b312 and left some comments about it

nkoenig commented:
> I tested this on the examples found in ign-gazebo. The only issue I found not mentioned as part of the known issues in #44 is the [diff_drive_skid](https://github.com/ignitionrobotics/ign-gazebo/blob/3e8d4e7dd97d6dccdab456625b4b17c9749ebf64/examples/worlds/diff_drive_skid.sdf) example. Commanding the vehicle to move forward in the `x` direction, with no rotation results in what's shown in the gif below:
> 
>  Peek.2021-05-11.17-49-skid.mp4 
> For comparison, this is the same example using dartsim
> 
>  Peek.2021-05-11.17-49-skid-dart.mp4

I captured this problem in #44.

azeey commented on bullet/src/KinematicsFeatures.cc:
```diff
@@ -0,0 +1,87 @@
+  btVector3 vel = rigidBody->getLinearVelocity();
+
+  // Transform to world frame
+  // const auto matBaseToWorld =
```
nit: Can the commented code be removed?

Lobotuerk replied:
371fbea should take care of this

azeey reviewed (approved):
Great work and thanks for iterating! Just a couple more nitpicks.

Lobotuerk commented on bullet/src/SDFFeatures_TEST.cc:
```diff
@@ -0,0 +1,147 @@
+    const auto &[model, joint] = CreateTestModel(world, "test1", parent, child);
+    EXPECT_NE(nullptr, joint);
+    }
+  */
```
371fbea should take care of this

Blast545 commented:
2 approvals, 10/10 checks passing, Merging
